common/djangoapps/third_party_auth/middleware.py

```diff
@@ -23,3 +25,46 @@ def get_redirect_uri(self, request, exception):
             redirect_uri = pipeline.AUTH_DISPATCH_URLS[auth_entry]
 
         return redirect_uri
+
+
+class ResetSessionIfPipelineBrokenMiddleware(object):
+    """
+    Middleware signs the user out if they need to provide data sharing consent,
+    haven't, and left the TPA pipeline prematurely
+    """
+    def process_view(self, request, view_func, view_args, view_kwargs):
+        """
+        Conditionally sign out users who don't provide data sharing consent
+        """
+        allowed_module_prefixes = (
```

Comment:
Hmmm. @mattdrayer, thoughts? I do think we want the middleware in some form; otherwise, we'd need a way to "suspend" the pipeline such that exiting it early doesn't keep an authenticated user connected to the session.